volatility/framework/plugins/mac/kevents.py

```diff
@@ -0,0 +1,185 @@
+        
+            klist_array = kernel.object(object_type = "array",
+                                    offset = klist_array_pointer,
+                                    count = array_size + 1,
+                                    subtype = kernel.get_type("klist"))
+        
+        except exceptions.InvalidAddressException:
+            return
+
+        for klist in klist_array:
+            for kn in mac.MacUtilities.walk_slist(klist, "kn_link"):
+                yield kn
+
+    @classmethod
+    def _get_task_kevents(cls, kernel, task):
+        """
+        Enumerates event filters per task.
+        Uses smear-safe APIs throughout as these data structures
+        see a signifcant amount of smear
+        """
+        fdp = task.p_fd
+
+        for kn in cls._walk_klist_array(kernel, fdp, "fd_knlist", "fd_knlistsize"):
+            yield kn
+
+        for kn in cls._walk_klist_array(kernel, fdp, "fd_knhash", "fd_knhashmask"):
+            yield kn
+
+        try:
+            p_klist = task.p_klist
+        except exceptions.InvalidAddressException:
+            return
+
+        for kn in mac.MacUtilities.walk_slist(p_klist, "kn_link"):
+            yield kn
+
+    @classmethod
+    def list_kernel_events(cls,
+                           context: interfaces.context.ContextInterface,
+                           layer_name: str,
+                           darwin_symbols: str,
+                           filter_func: Callable[[int], bool] = lambda _: False) -> \
+            Iterable[Tuple[interfaces.objects.ObjectInterface,
+                           interfaces.objects.ObjectInterface,
+                           interfaces.objects.ObjectInterface]]:
+        """
+        Returns the kernel event filters registered
+
+        Return values:
+            A tuple of 3 elements:
+                1) The name of the process that registered the filter
+                2) The process ID of the process that registered the filter
+                3) The object of the associated kernel event filter
+        """
+        kernel = contexts.Module(context, darwin_symbols, layer_name, 0)
+       
+        list_tasks = pslist.PsList.get_list_tasks(pslist.PsList.pslist_methods[0])
+        
+        for task in list_tasks(context, layer_name, darwin_symbols, filter_func):
+            task_name = utility.array_to_string(task.p_comm)
+            pid = task.p_pid
+            
+            for kn in cls._get_task_kevents(kernel, task):
+                yield task_name, pid, kn
+
+    def _generator(self):
+        filter_func = pslist.PsList.create_pid_filter(self.config.get('pid', None))
+
+        for task_name, pid, kn in self.list_kernel_events(self.context,
+                                                self.config['primary'],
+                                                self.config['darwin'],
+                                                filter_func = filter_func):
+
+            filter_index = kn.kn_kevent.filter * -1
```

Comment:
This seems a bit odd, is it just flipping the sign bit (should the type really be unsigned) or is something else going on?

Reply:
it is actually stored as a negative number in the kernel.

Open this:

https://opensource.apple.com/source/xnu/xnu-517.9.5/bsd/sys/event.h.auto.html

then search:

EVFILT_READ

and you will see they are all negative numbers, it is strange.